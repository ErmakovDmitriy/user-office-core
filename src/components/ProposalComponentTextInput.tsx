import React, { ChangeEvent } from "react";
import { TextField, makeStyles } from "@material-ui/core";
import { IBasicComponentProps } from "./IBasicComponentProps";
export function ProposalComponentTextInput(props: IBasicComponentProps) {
  const classes = makeStyles({
    textField: {
      margin: "15px 0 10px 0"
    }
  })();
  let { templateField, onComplete, touched, errors, handleChange } = props;
  let { proposal_question_id, config, question } = templateField;
  let isError = (touched[proposal_question_id] && errors[proposal_question_id]) ? true : false;
  return (
    <div>
      <TextField
        id={proposal_question_id}
        name={proposal_question_id}
        fullWidth
        label={question}
        value={templateField.value}
        onChange={(evt: ChangeEvent<HTMLInputElement>) => {
          templateField.value = evt.target.value;
          handleChange(evt); // letting Formik know that there was a change
        }}
        onBlur={() => onComplete()}
        placeholder={config.small_label}
        error={isError}
        helperText={errors[proposal_question_id]}
        multiline={config.multiline}
        rows={config.multiline ? 4 : 1}
        rowsMax={config.multiline ? 16 : undefined}
        className={classes.textField}
        InputLabelProps={{
          shrink: true
        }}
      />
    </div>
  );
}
