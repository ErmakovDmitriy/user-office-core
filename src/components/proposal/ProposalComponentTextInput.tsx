import { makeStyles } from '@material-ui/core';
import { getIn } from 'formik';
import React, { ChangeEvent, useState, useEffect } from 'react';

import { TextInputConfig } from '../../generated/sdk';
import TextFieldWithCounter from '../common/TextFieldWithCounter';
import { BasicComponentProps } from './IBasicComponentProps';

export function ProposalComponentTextInput(props: BasicComponentProps) {
  const classes = makeStyles({
    textField: {
      margin: '15px 0 10px 0',
    },
  })();
  const { templateField, touched, errors, onComplete } = props;
  const {
    question: { proposalQuestionId },
    question,
    value,
  } = templateField;
  const [stateValue, setStateValue] = useState(value);
  const fieldError = getIn(errors, proposalQuestionId);
  const isError = getIn(touched, proposalQuestionId) && !!fieldError;
  const config = templateField.question.config as TextInputConfig;

  useEffect(() => {
    setStateValue(templateField.value);
  }, [templateField]);

  return (
    <div>
      {config.htmlQuestion && (
        <div
          dangerouslySetInnerHTML={{
            __html: config.htmlQuestion!,
          }}
        ></div>
      )}
      <TextFieldWithCounter
        variant="standard"
        id={proposalQuestionId}
        name={proposalQuestionId}
        fullWidth
        required={config.required ? true : false}
        label={config.htmlQuestion ? '' : question.question}
        value={stateValue}
        onChange={(evt: ChangeEvent<HTMLInputElement>) => {
          setStateValue(evt.target.value);
        }}
        onBlur={evt => {
          onComplete(evt, evt.target.value);
        }}
        placeholder={config.placeholder}
        error={isError}
        helperText={isError && errors[proposalQuestionId]}
        multiline={config.multiline}
        rows={config.multiline ? 2 : 1}
        rowsMax={config.multiline ? 16 : undefined}
        className={classes.textField}
        InputLabelProps={{
          shrink: true,
        }}
        maxLen={config.max || undefined}
      />
    </div>
  );
}
