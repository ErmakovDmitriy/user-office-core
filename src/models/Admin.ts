import { ObjectType, Field, Int } from "type-graphql";

@ObjectType()
export class Page {
  @Field(() => Int)
  public id: number;

  @Field({ nullable: true })
  public content: string;

  constructor(id: number, content: string) {
    this.id = id;
    this.content = content;
  }
}
