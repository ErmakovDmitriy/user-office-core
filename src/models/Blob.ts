export class FileMetaData {
    constructor(
        public file_id:string,
        public oid:number,
        public originalFileName:string,
        public mimeType:string,
        public sizeInBytes:number,
        public createdDate:Date
    ) {}
}