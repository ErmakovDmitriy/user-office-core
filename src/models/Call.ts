export class Call {
  constructor(
    public id: number,
    public shortCode: string,
    public startCall: Date,
    public endCall: Date,
    public startReview: Date,
    public endReview: Date,
    public startNotify: Date,
    public endNotify: Date,
    public startCycle: Date,
    public endCycle: Date,
    public cycleComment: string,
    public surveyComment: string,
    public templateId?: number
  ) {}
}
