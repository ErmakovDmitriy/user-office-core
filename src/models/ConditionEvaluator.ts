import JSDict from "../utils/Dictionary";
import { EvaluatorOperator, QuestionaryField } from "../generated/sdk";

export class ConditionEvaluator {
  private validatorMap!: JSDict<EvaluatorOperator, IFieldConditionEvaluator>;

  private getMappings() {
    if (!this.validatorMap) {
      // lazy initialization
      this.validatorMap = JSDict.Create<
        EvaluatorOperator,
        IFieldConditionEvaluator
      >();
      this.validatorMap.put(EvaluatorOperator.EQ, new EqualityValidator());
      this.validatorMap.put(EvaluatorOperator.NEQ, new InequalityValidator());
    }
    return this.validatorMap;
  }

  getConditionEvaluator(id: EvaluatorOperator): IFieldConditionEvaluator {
    return this.getMappings().get(id)!;
  }
}

export interface IFieldConditionEvaluator {
  isSatisfied(field: QuestionaryField, params: string): boolean;
}

export class EqualityValidator implements IFieldConditionEvaluator {
  isSatisfied(field: QuestionaryField, params: string): boolean {
    return field.value === params;
  }
}

export class InequalityValidator implements IFieldConditionEvaluator {
  isSatisfied(field: QuestionaryField, params: string): boolean {
    return field.value !== params;
  }
}
