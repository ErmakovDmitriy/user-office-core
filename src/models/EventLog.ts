export class EventLog {
  constructor(
    public id: number,
    public changedBy: number,
    public eventType: string,
    public rowData: string,
    public eventTStamp: string,
    public changedObjectId: number
  ) {}
}
