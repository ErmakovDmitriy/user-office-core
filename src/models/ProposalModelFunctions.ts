import { ConditionEvaluator } from "./ConditionEvaluator";
import {
  ProposalTemplateField,
  QuestionaryField,
  ProposalTemplate,
  Questionary,
  DataType,
  DataTypeSpec,
  FieldDependency,
  FieldConfig,
  TextInputConfig,
  SelectionFromOptionsConfig
} from "./ProposalModel";
import JSDict from "../utils/Dictionary";
type AbstractField = ProposalTemplateField | QuestionaryField;
type AbstractCollection = ProposalTemplate | Questionary;
export function getDataTypeSpec(type: DataType): DataTypeSpec {
  switch (type) {
    case DataType.EMBELLISHMENT:
      return { readonly: true };
    default:
      return { readonly: false };
  }
}
export function getTopicById(collection: AbstractCollection, topicId: number) {
  const step = collection.steps.find(step => step.topic.topic_id === topicId);
  return step ? step.topic : undefined;
}
export function getQuestionaryStepByTopicId(
  template: AbstractCollection,
  topicId: number
) {
  return template.steps.find(step => step.topic.topic_id === topicId);
}
export function getFieldById(
  collection: AbstractCollection,
  questionId: string
) {
  let needle: AbstractField | undefined;
  collection.steps.every(step => {
    needle = step.fields.find(
      field => field.proposal_question_id === questionId
    );
    return needle === undefined;
  });
  return needle;
}
export function getAllFields(collection: AbstractCollection) {
  let allFields = new Array<AbstractField>();
  collection.steps.forEach(step => {
    allFields = allFields.concat(step.fields);
  });
  return allFields;
}
export function isDependencySatisfied(
  collection: Questionary,
  dependency: FieldDependency
): boolean {
  //const { condition, params } = dependency.condition;
  const { condition, params } = JSON.parse(dependency.condition); // TODO SWAP-341. strongly type this after making GraphQL able to return more custom objects
  const field = getFieldById(
    collection,
    dependency.proposal_question_dependency
  ) as QuestionaryField | undefined;
  if (!field) {
    return true;
  }
  const isParentSattisfied = areDependenciesSatisfied(
    collection,
    dependency.proposal_question_dependency
  );
  return (
    isParentSattisfied &&
    new ConditionEvaluator()
      .getConfitionEvaluator(condition)
      .isSattisfied(field, params)
  );
}
export function areDependenciesSatisfied(
  questionary: Questionary,
  fieldId: string
) {
  const field = getFieldById(questionary, fieldId);
  if (!field) {
    return true;
  }
  const isAtLeastOneDissasisfied = field.dependencies!.some(dep => {
    let result = isDependencySatisfied(questionary, dep) === false;
    return result;
  });
  return isAtLeastOneDissasisfied === false;
}

export function isMatchingConstraints(
  value: any,
  field: ProposalTemplateField
): boolean {
  const val = JSON.parse(value).value;
  const validator = validatorMap.get(field.data_type) || new BaseValidator();
  return validator.validate(val, field);
}

interface IConstraintValidator {
  validate(value: any, field: ProposalTemplateField): boolean;
}

class BaseValidator implements IConstraintValidator {
  constructor(private dataType?: DataType | undefined) {}

  validate(value: any, field: QuestionaryField) {
    if (this.dataType && field.data_type !== this.dataType) {
      throw new Error("Field validator ");
    }
    if (field.config.required && !value) {
      return false;
    }
    return true;
  }
}

class TextInputValidator extends BaseValidator {
  constructor() {
    super(DataType.TEXT_INPUT);
  }
  validate(value: any, field: QuestionaryField) {
    if (!super.validate(value, field)) {
      return false;
    }
    const config = field.config as TextInputConfig;
    if (config.min && value && value.length < config.min) {
      return false;
    }
    if (config.max && value && value.length > config.max) {
      return false;
    }
    return true;
  }
}

class SelectFromOptionsInputValidator extends BaseValidator {
  constructor() {
    super(DataType.SELECTION_FROM_OPTIONS);
  }
  validate(value: any, field: QuestionaryField) {
    const config = field.config as SelectionFromOptionsConfig;
    if (!super.validate(value, field)) {
      return false;
    }

    if (config.required && config.options!.indexOf(value) === -1) {
      return false;
    }

    return true;
  }
}

const validatorMap = JSDict.Create<DataType, IConstraintValidator>();
validatorMap.put(DataType.TEXT_INPUT, new TextInputValidator());
validatorMap.put(
  DataType.SELECTION_FROM_OPTIONS,
  new SelectFromOptionsInputValidator()
);
