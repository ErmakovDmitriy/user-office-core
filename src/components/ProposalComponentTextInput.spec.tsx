import React from "react";
import ReactTestRenderer from "react-test-renderer";
import { ProposalComponentTextInput } from "./ProposalComponentTextInput";
import { create1Topic3FieldWithDependenciesQuestionary } from "../tests/ProposalTestBed";
import { getFieldById } from "../models/ProposalModelFunctions";
import { QuestionaryField } from "../generated/sdk";

test("TextField rendered without crashing", () => {
  var template = create1Topic3FieldWithDependenciesQuestionary();
  const renderer = ReactTestRenderer.create(
    <ProposalComponentTextInput
      templateField={
        getFieldById(template, "links_with_industry")! as QuestionaryField
      }
      onComplete={() => {}}
      handleChange={() => {}}
      errors={{}}
      touched={{}}
    />
  );

  let tree = renderer.toJSON();
  expect(tree).toMatchSnapshot();
});
