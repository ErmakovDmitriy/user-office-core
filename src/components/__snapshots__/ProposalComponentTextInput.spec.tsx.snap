// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textfield rendered without crashing 1`] = `
<div>
  <div
    className="MuiFormControl-root MuiTextField-root makeStyles-textField-1  MuiFormControl-fullWidth"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
      data-shrink={true}
      htmlFor="links_with_industry"
    >
      If yes, please describe:
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        className="MuiInputBase-input MuiInput-input"
        disabled={false}
        id="links_with_industry"
        name="links_with_industry"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Please specify links with industry"
        required={false}
        rows={1}
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
